flink-ml-lib/src/main/java/org/apache/flink/ml/common/linalg/DenseMatrix.java

```diff
@@ -0,0 +1,577 @@
+        for (int i = 0; i < sub.m; i++) {
+            for (int j = 0; j < sub.n; j++) {
+                this.set(m0 + i, n0 + j, sub.get(i, j));
+            }
+        }
+    }
+
+    /**
+     * Set a single element.
+     *
+     * @param i Row index.
+     * @param j Column index.
+     * @param s A(i,j).
+     * @throws ArrayIndexOutOfBoundsException
+     */
+    public void set(int i, int j, double s) {
+        data[j * m + i] = s;
+    }
+
+    /**
+     * Add the given value to a single element.
+     *
+     * @param i Row index.
+     * @param j Column index.
+     * @param s A(i,j).
+     * @throws ArrayIndexOutOfBoundsException
+     */
+    public void add(int i, int j, double s) {
+        data[j * m + i] += s;
+    }
+
+    /**
+     * Check whether the matrix is square matrix.
+     *
+     * @return <code>true</code> if this matrix is a square matrix, <code>false</code> otherwise.
+     */
+    public boolean isSquare() {
+        return m == n;
+    }
+
+    /**
+     * Check whether the matrix is symmetric matrix.
+     *
+     * @return <code>true</code> if this matrix is a symmetric matrix, <code>false</code> otherwise.
+     */
+    public boolean isSymmetric() {
+        if (m != n) {
+            return false;
+        }
+        for (int i = 0; i < n; i++) {
+            for (int j = i + 1; j < n; j++) {
+                if (this.get(i, j) != this.get(j, i)) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Get the number of rows.
+     *
+     * @return the number of rows.
+     */
+    public int numRows() {
+        return m;
+    }
+
+    /**
+     * Get the number of columns.
+     *
+     * @return the number of columns.
+     */
+    public int numCols() {
+        return n;
+    }
+
+    /** Sum of all elements of the matrix. */
+    public double sum() {
+        double s = 0.;
+        for (int i = 0; i < this.data.length; i++) {
+            s += this.data[i];
+        }
+        return s;
+    }
+
+    /** Scale the vector by value "v" and create a new matrix to store the result. */
+    public DenseMatrix scale(double v) {
+        DenseMatrix r = this.clone();
+        BLAS.scal(v, r);
+        return r;
+    }
+
+    /** Scale the matrix by value "v". */
+    public void scaleEqual(double v) {
+        BLAS.scal(v, this);
+    }
+
+    /** Create a new matrix by plussing another matrix. */
+    public DenseMatrix plus(DenseMatrix mat) {
+        DenseMatrix r = this.clone();
+        BLAS.axpy(1.0, mat, r);
+        return r;
+    }
+
+    /** Create a new matrix by plussing a constant. */
+    public DenseMatrix plus(double alpha) {
+        DenseMatrix r = this.clone();
+        for (int i = 0; i < r.data.length; i++) {
+            r.data[i] += alpha;
+        }
+        return r;
+    }
+
+    /** Plus with another matrix. */
+    public void plusEquals(DenseMatrix mat) {
```

Comment:
We will need to discuss the API of these Flink ML infra classes. There are in general two styles. The approach taken in this PR adds quite a few methods here. Another approach is to keep only the very basic operations on the DenseMatrix API and move computation methods to a dedicated class (e.g. BLAS). This is pretty much what Spark does.

It is probably good to start simple and keep only the basic API needed by the algorithms we planned for FFA. We can discuss offline on this as well.

Reply:
OK